### PR TITLE
[Integration][AWS] fix: Cannot read global resources without regional permissions

### PR DESCRIPTION
**Pull request by shalev007**
- **feat: add access denied handler to STS:AssumeRole**
- **feat: add access denied handler to custom kind resync**
- **feat: brute force global resource if does not have permissions in default region**

# Description

What - Read all regions in global resource
Why - Whenever there are no permissions in a region global resources (such as AWS::S3::Bucket) that only use the first default region cannot read any resources.
How - Read all regions until there's resources in any region

## Type of change

Please leave one option from the following and delete the rest:

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

### CODE REVIEW

File: integrations/aws/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+# Port_Ocean 0.2.16 (2024-07-16)
+
+### Improvements
+
+- Brute force global resource if does not have permissions in default region (#1)
```
Comment (Tankilevitch) [inline integrations/aws/CHANGELOG.md:14]: please use more appropriate explanation about what and why we chose to implement that

**Review by erikzaadi — APPROVED** [review]
Absolute joy to go over the split of commits, logical change per commit, super easy to review.
Kudos!!1

File: integrations/aws/aws/session_manager.py [hunk 2]
```diff
@@ -90,16 +90,25 @@ async def _get_organization_session(self) -> aioboto3.Session | None:
...
+                organization_role_session = aioboto3.Session(
+                    aws_access_key_id=credentials["AccessKeyId"],
+                    aws_secret_access_key=credentials["SecretAccessKey"],
+                    aws_session_token=credentials["SessionToken"],
+                )
+            except sts_client.exceptions.ClientError as e:
+                if e.response.get("Error", {}).get("Code") == "AccessDenied":
+                    logger.warning(
+                        "Caller does not have permission to assume the organization role. Assuming application role has access to organization accounts."
+                    )
```
Comment (Tankilevitch) [inline integrations/aws/aws/session_manager.py:108]: what does it mean 😅  leave a comment about the decision of this implementation

File: integrations/aws/utils/resources.py [hunk 3]
```diff
@@ -128,32 +128,42 @@ async def resync_custom_kind(
...
+                if results:
+                    yield [
+                        {
+                            CustomProperties.KIND: kind,
+                            CustomProperties.ACCOUNT_ID: account_id,
+                            CustomProperties.REGION: region,
+                            **fix_unserializable_date_properties(resource),
+                        }
+                        for resource in results
+                    ]
```
Comment (Tankilevitch) [inline integrations/aws/utils/resources.py:150]: lets log also that we received empty results, same as my previous pr for regular kinds

File: integrations/aws/utils/resources.py [hunk 4]
```diff
@@ -222,9 +232,18 @@ async def resync_cloudcontrol(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
                         == "AccessDeniedException"
                     ):
+                        if is_global:
+                            logger.warning(
+                                f"Access denied in global resource {kind} in region {region}. Querying again in all regions"
+                            )
+                            async for batch in resync_cloudcontrol(
+                                kind, is_global=False
+                            ):
+                                yield batch
```
Comment (Tankilevitch) [inline integrations/aws/utils/resources.py:242]: add message about the why and the intention

File: integrations/aws/utils/resources.py [hunk 5]
```diff
@@ -222,9 +232,18 @@ async def resync_cloudcontrol(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
+                                return  # stop once we have some resources
                         logger.warning(
                             f"Skipping resyncing {kind} in region {region} due to missing access permissions"
                         )
-                        break
+                        break  # stop pagination if access is denied in this region
```
Comment (Tankilevitch) [inline integrations/aws/utils/resources.py:247]: I miss the why

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Nice job, left some comments

File: integrations/aws/utils/resources.py (hunk 4 above)
Comment (shalev007) [reply integrations/aws/utils/resources.py:242]: I did a comment explaining why, but then decided to leave it as as a log to inform the user as well, check out the log above, doesn't it suffice the "Why"?

File: integrations/aws/aws/session_manager.py (hunk 2 above)
Comment (shalev007) [reply integrations/aws/aws/session_manager.py:108]: rephrased it, please have a look

File: integrations/aws/utils/resources.py (hunk 3 above)
Comment (shalev007) [reply integrations/aws/utils/resources.py:150]: There's already a log to indicate the amount of resources received, Do you think we should also add log to indicate empty results specifically?

File: integrations/aws/utils/resources.py (hunk 5 above)
Comment (shalev007) [reply integrations/aws/utils/resources.py:247]: rephrased it, please have a look

File: integrations/aws/CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 <!-- towncrier release notes start -->
 
+# Port_Ocean 0.2.16 (2024-07-16)
+
+### Improvements
```
Comment (Tankilevitch) [inline integrations/aws/CHANGELOG.md:12]:
```suggestion

### Bug Fixes
```

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: integrations/aws/utils/resources.py [hunk 7]
```diff
@@ -222,9 +232,18 @@ async def resync_cloudcontrol(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
                         e.response.get("Error", {}).get("Code")
                         == "AccessDeniedException"
                     ):
+                        if is_global:
+                            logger.warning(
+                                f"Access denied in global resource {kind} in region {region}. Querying again in all regions"
+                            )
+                            async for batch in resync_cloudcontrol(
+                                kind, is_global=False
```
Comment (Tankilevitch) [inline integrations/aws/utils/resources.py:240]: if I dont have access to region x which is the default, but we do have access to region y, z will we query both y and z?

Comment (shalev007) [reply integrations/aws/utils/resources.py:240]: Yes, until the query will return some data or we run on all regions

File: integrations/aws/aws/session_manager.py [hunk 8]
```diff
@@ -153,7 +161,7 @@ async def _assume_role_and_update_credentials(
             self._aws_credentials.append(credentials)
             self._aws_accessible_accounts.append(account)
         except sts_client.exceptions.ClientError as e:
-            if e.response["Error"]["Code"] == "AccessDenied":
+            if is_access_denied_exception(e):
```
Comment (Tankilevitch) [inline integrations/aws/aws/session_manager.py:164]: ❤️

File: integrations/aws/main.py [hunk 9]
```diff
@@ -36,8 +38,16 @@ async def resync_all(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
-    async for batch in resync_cloudcontrol(kind):
-        yield batch
+    is_global = is_global_resource(kind)
+    try:
+        async for batch in resync_cloudcontrol(kind, is_global):
+            yield batch
+    except Exception as e:
+        if is_access_denied_exception(e):
+            async for batch in resync_cloudcontrol(
+                kind, is_global=False, stop_on_first_data=True
```
Comment (Tankilevitch) [inline integrations/aws/main.py:48]:
```suggestion
                kind, is_global=False, stop_on_first_region=True
```

File: integrations/aws/aws/session_manager.py [hunk 10]
```diff
@@ -90,16 +91,23 @@ async def _get_organization_session(self) -> aioboto3.Session | None:
...
+                credentials = organizations_client["Credentials"]
+                organization_role_session = aioboto3.Session(
+                    aws_access_key_id=credentials["AccessKeyId"],
+                    aws_secret_access_key=credentials["SecretAccessKey"],
+                    aws_session_token=credentials["SessionToken"],
+                )
+            except sts_client.exceptions.ClientError as e:
+                if is_access_denied_exception(e):
+                    logger.warning("Cannot assume role to the organization account.")
+                    return self._application_session
```
Comment (Tankilevitch) [inline integrations/aws/aws/session_manager.py:108]: Add comment for why we do that, and log that we are using application client instead

**Pull request merged**